### ISSUE TITLE
Pyright only relies on first definition of Union of Callable for lambdas

### ISSUE THREAD

**Issue by NCPlayz** [microsoft/pyright#2222]
**Describe the bug**
Guess there's two parts to this.
When you accept a union of callables with a different length of arguments, pyright seems to only use the first definition when inferring the types for lambda arguments. This causes two issues:

 - Error 1: It relies on `Callable[[int, str], bool]`, which infers that `s` is `int`, and that it expected one more argument.
 - Error 2: It relies on `Callable[[str], bool]`, which infers that `i` is `str`, and `s` is `Unknown`. 

**To Reproduce**
See code.

**Expected behavior**
Both errors should not be raised, and it should be able to correctly infer the arguments, rather than only reflect on the first possibility of the Union.

**Screenshots or Code**
```py
from typing import Callable, Union


Check = Union[Callable[[int, str], bool], Callable[[str], bool]]
WorksWithLambdaCheck = Union[Callable[[str], bool], Callable[[int, str], bool]]

def check(check: Check) -> Check:
    return check

def check_lambda(check: WorksWithLambdaCheck) -> WorksWithLambdaCheck:
    return check

def check_callback_1(integer: int, string: str):
    return True

def check_callback_2(string: str):
    return True

check(lambda s: s.startswith("hello")) # error 1
check(lambda i, s: i > 0 and s.startswith("hello"))
check(check_callback_1)
check(check_callback_2)

check_lambda(lambda s: s.startswith("hello"))
check_lambda(lambda i, s: i > 0 and s.startswith("hello")) # error 2
check_lambda(check_callback_1)
check_lambda(check_callback_2)
```

**VS Code extension or command-line**
VS Code - v2021.8.2

**Additional context**
Here are the full errors returned by Pyright:

Error 1:

```py
Argument of type "(s: int) -> Unknown" cannot be assigned to parameter "check" of type "Check" in function "check"
  Type "(s: int) -> Unknown" cannot be assigned to type "Check"
    Type "(s: int) -> Unknown" cannot be assigned to type "(_p0: int, _p1: str) -> bool"
      Function accepts too many positional parameters; expected 1 but received 2
    Type "(s: int) -> Unknown" cannot be assigned to type "(_p0: str) -> bool"
      Parameter 1: type "str" cannot be assigned to type "int"
        "str" is incompatible with "int" Pylance(reportGeneralTypeIssues)
```

Error 2:

```py
Argument of type "(i: str, s: Unknown) -> Unknown" cannot be assigned to parameter "check" of type "WorksWithLambdaCheck" in function "check_lambda"
  Type "(i: str, s: Unknown) -> Unknown" cannot be assigned to type "WorksWithLambdaCheck"
    Type "(i: str, s: Unknown) -> Unknown" cannot be assigned to type "(_p0: str) -> bool"
      Keyword parameter "s" is missing in destination
    Type "(i: str, s: Unknown) -> Unknown" cannot be assigned to type "(_p0: int, _p1: str) -> bool"
      Parameter 1: type "int" cannot be assigned to type "str"
        "int" is incompatible with "str" Pylance(reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
Thanks for the bug report. Yes, this is a current limitation in pyright's logic. I will work on an improvement. At the very least, it should consider only those callable subtypes whose arity (parameter count) matches that of the lambda.

**Comment by erictraut** [comment]
This will be addressed in the next release. The logic now takes into account the number of parameters in each of the union subtypes.

**Comment by erictraut** [comment]
This is included in pyright 1.1.164, which I just published. It will also be in the next release of Pylance.
